util/dvsim/Modes.py

```diff
@@ -84,7 +84,7 @@ def merge_mode(self, mode):
             self_attr_val = getattr(self, attr)
             mode_attr_val = getattr(mode, attr)
 
-            if type(self_attr_val) is list:
+            if type(self_attr_val) is list and type(mode_attr_val) is list:
```

Comment:
If I understand this patch right, the point is that `mode_attr_val` might now be None. I had to squint quite hard to understand what all the types meant, and some comments would make it much easier to read. Can we change this to something more like this?

```python
        # Try to merge each of our attributes in self with the value from
        # mode.
        for attr, self_attr_val in self.__dict__.items():
            mode_attr_val = getattr(mode, attr, None)

            # As a special case, sub-modes can have different names. Skip any
            # checks for name/mname in this case.
            if is_sub_mode and attr in ['name', 'mname']:
                continue

            # If our attribute is a list and mode_attr_val is specified, it
            # should also be a list and we append it.
            if isinstance(self_attr_val, list) and mode_attr_val is not None:
                assert isinstance(mode_attr_val, list)
                self_attr_val.extend(mode_attr_val)
                continue

            # If our attribute isn't a list, it's some atomic type (int or
            # str). Check that mode_attr_val is compatible.
            self.check_conflict(mode.name, attr, mode_attr_val)
```

As well as the comments, this adds an assertion that the types make sense in the "extend a list" case, so things will explode if `mode_attr_val` isn't either a list or None.

It also gets rid of an unnecessary `setattr`: the `list.extend` function modifies `self`, so `self_attr_val.extend(mode_attr_val)` already updates the attribute.

Reply:
I second the need for comments to create clearity